program/rust/src/processor/upd_price.rs

```diff
@@ -161,37 +162,76 @@ pub fn upd_price(
-        };
-
-        // If the aggregate was successfully updated, calculate the difference and update TWAP.
-        if updated {
-            let agg_diff = (clock.slot as i64)
-                - load_checked::<PriceAccount>(price_account, cmd_args.header.version)?.prev_slot_
-                    as i64;
-            // Encapsulate TWAP update logic in a function to minimize unsafe block scope.
-            unsafe {
-                c_upd_twap(price_account.try_borrow_mut_data()?.as_mut_ptr(), agg_diff);
-            }
+
+    // Try to update the publisher's price
+    if is_component_update(cmd_args)? {
+        // IMPORTANT: If the publisher does not meet the price/conf
+        // ratio condition, its price will not count for the next
+        // aggregate.
+        let status: u32 =
+            get_status_for_conf_price_ratio(cmd_args.price, cmd_args.confidence, cmd_args.status)?;
+
+        {
+            let mut price_data =
+                load_checked::<PriceAccount>(price_account, cmd_args.header.version)?;
+            let publisher_price = &mut price_data.comp_[publisher_index].latest_;
+            publisher_price.price_ = cmd_args.price;
+            publisher_price.conf_ = cmd_args.confidence;
+            publisher_price.status_ = status;
+            publisher_price.pub_slot_ = cmd_args.publishing_slot;
+        }
+    }
+
+    // get number of slots from last update
+    let slots_since_last_update = clock.slot - latest_aggregate_price.pub_slot_;
+
+    // If the price update is the first in the slot and the aggregate is trading, update the previous slot, price, conf, and timestamp.
+    if slots_since_last_update > 0 && latest_aggregate_price.status_ == PC_STATUS_TRADING {
+        let mut price_data = load_checked::<PriceAccount>(price_account, cmd_args.header.version)?;
+        price_data.prev_slot_ = price_data.agg_.pub_slot_;
+        price_data.prev_price_ = price_data.agg_.price_;
+        price_data.prev_conf_ = price_data.agg_.conf_;
+        price_data.prev_timestamp_ = clock.unix_timestamp;
+    };
+
+
+    let updated = unsafe {
+        // NOTE: c_upd_aggregate must use a raw pointer to price
+        // data. Solana's `<account>.borrow_*` methods require exclusive
+        // access, i.e. no other borrow can exist for the account.
+        c_upd_aggregate(
+            price_account.try_borrow_mut_data()?.as_mut_ptr(),
+            clock.slot,
+            clock.unix_timestamp,
+        )
+    };
+
+    // If the aggregate was successfully updated, calculate the difference and update TWAP.
+    if updated {
+        let agg_diff = (clock.slot as i64)
+            - load_checked::<PriceAccount>(price_account, cmd_args.header.version)?.prev_slot_
+                as i64;
+        {
             let mut price_data =
                 load_checked::<PriceAccount>(price_account, cmd_args.header.version)?;
+            // check if its the first price update in the slot
+            if slots_since_last_update > 0 {
```

Comment:
I'm not sure I agree about this one

Reply:
> I think this if statement is wrong. What if you have a slot where the first aggregate update sets the status to UNKNOWN, then a later update sets it to TRADING? In that case, this if statement will return false and we will forget to update the twap.
> 
> I think the right check here is to explicitly compare the clock slot with a stored last updated slot for the prev_twap_ (etc) fields.

I'm not sure I get this, `c_upd_aggregate` will only return `True` if agg price is successfully updated and `ptr->agg_.status_` will be `PC_STATUS_TRADING`, and in that case this block will only be ran if `ptr->agg_.status_ = PC_STATUS_TRADING`

Reply:
changed slots_since_last_update to be `clock.slot - price_data.last_slot_` instead since `last_slot` will only be updated if the agg price is successfully updated